File: apps/Image-to-3D/iOS/ImageTo3D/Model/DepthModel.swift
import Foundation
import ZeticMLange

/// Thin wrapper over ZeticMLangeModel: loads once, runs depth inference, times latency.
final class DepthModel {
    private var model: ZeticMLangeModel?
    private(set) var lastInferMs: Double = 0

    enum ModelError: Error, LocalizedError {
        case notLoaded, emptyOutput

        var errorDescription: String? {
            switch self {
            case .notLoaded: return "Model is not loaded yet"
            case .emptyOutput: return "Model returned no output tensors"
            }
        }
    }

    /// Loads/downloads the model on a background queue. Callbacks fire on the main thread.
    func load(onProgress: @escaping (Float) -> Void,
              completion: @escaping (Result<Void, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let m = try ZeticMLangeModel(
                    personalKey: AppConfig.personalKey,
                    name: AppConfig.modelName,
                    version: AppConfig.modelVersion,   // nil = latest
                    modelMode: .RUN_AUTO,
                    onDownload: { progress in
                        DispatchQueue.main.async { onProgress(progress) }
                    }
                )
                self.model = m
                print("[ImageTo3D] model loaded: \(AppConfig.modelName)")
                DispatchQueue.main.async { completion(.success(())) }
            } catch {
                print("[ImageTo3D] model load FAILED: \(error)")
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    /// Runs depth inference on a CHW planar [0, 1] RGB buffer.
    func infer(_ chw: [Float]) throws -> DepthMap {
        guard let model else { throw ModelError.notLoaded }
        let size = AppConfig.inputSize

        let data = chw.withUnsafeBufferPointer { Data(buffer: $0) }
        let input = ZeticMLange.Tensor(data: data,
                                       dataType: ZeticMLange.BuiltinDataType.float32,
                                       shape: [1, 3, size, size])

        let t0 = CFAbsoluteTimeGetCurrent()
        let outputs = try model.run(inputs: [input])
        lastInferMs = (CFAbsoluteTimeGetCurrent() - t0) * 1000

        guard let first = outputs.first else { throw ModelError.emptyOutput }
        return try DepthMap(tensor: first)
    }
}

File: apps/Image-to-3D/Android/app/src/main/java/ai/zetic/demo/imageto3d/DepthModel.kt
package ai.zetic.demo.imageto3d

import android.content.Context
import android.util.Log
import com.zeticai.mlange.core.model.ModelMode
import com.zeticai.mlange.core.model.QuantType
import com.zeticai.mlange.core.model.ZeticMLangeModel
import com.zeticai.mlange.core.tensor.DataType
import com.zeticai.mlange.core.tensor.Tensor
import java.nio.ByteBuffer
import java.nio.ByteOrder

/** Depth map decoded from the model's output tensor (relative inverse depth,
 *  larger = closer), with a robust 2/98-percentile range. */
class DepthMap(val values: FloatArray, val width: Int, val height: Int) {
    val robustLo: Float
    val robustHi: Float

    init {
        val sample = FloatArray((values.size + 15) / 16)
        var j = 0
        var i = 0
        while (i < values.size) {
            sample[j++] = values[i]
            i += 16
        }
        sample.sort()
        val lo = sample[((sample.size - 1) * 0.02f).toInt()]
        val hi = sample[((sample.size - 1) * 0.98f).toInt()]
        robustLo = lo
        robustHi = maxOf(hi, lo + 1e-6f)
    }

    /** Disparity normalized to [0, 1] over the robust range; 1 = closest. */
    fun normalized(index: Int): Float =
        ((values[index] - robustLo) / (robustHi - robustLo)).coerceIn(0f, 1f)
}

/** Thin wrapper over ZeticMLangeModel: loads once, runs inference, times latency. */
class DepthModel {
    private var model: ZeticMLangeModel? = null
    var lastInferMs = 0.0
        private set

    /** Blocking; call from a background thread. */
    fun load(context: Context, onProgress: (Float) -> Unit) {
        model = ZeticMLangeModel(
            context,
            AppConfig.PERSONAL_KEY,
            AppConfig.MODEL_NAME,
            null,                 // version: latest
            ModelMode.RUN_AUTO,
            QuantType.FP32,
            { p: Float -> onProgress(p) },
        )
        Log.i(TAG, "model loaded: ${AppConfig.MODEL_NAME}")
    }

    fun infer(chw: FloatArray): DepthMap {
        val model = model ?: error("model not loaded")
        val size = AppConfig.INPUT_SIZE

        val buffer = ByteBuffer.allocateDirect(chw.size * 4).order(ByteOrder.nativeOrder())
        buffer.asFloatBuffer().put(chw)
        buffer.rewind()
        val input = Tensor(buffer, DataType.Float32, intArrayOf(1, 3, size, size))

        val t0 = System.nanoTime()
        val outputs = model.run(arrayOf(input))
        lastInferMs = (System.nanoTime() - t0) / 1e6

        val out = outputs.firstOrNull() ?: error("model returned no outputs")
        val bb = out.data.order(ByteOrder.nativeOrder())
        bb.rewind()
        val floats = FloatArray(bb.remaining() / 4)
        bb.asFloatBuffer().get(floats)

        // Defensive shape handling: expect H*W floats; sqrt fallback otherwise.
        val expected = size * size
        val (w, h) = if (floats.size == expected) {
            size to size
        } else {
            val s = Math.sqrt(floats.size.toDouble()).toInt()
            check(s * s == floats.size) { "unexpected output count ${floats.size}" }
            s to s
        }
        return DepthMap(floats, w, h)
    }

    fun deinit() {
        model?.close()
        model = null
    }

    private companion object { const val TAG = "ImageTo3D" }
}
